### PR TITLE
PLFM-9632 Add IdP to access token

### PR DESCRIPTION
**Pull request by brucehoff**
This PR adds the Identity Provider (IdP) used to authenticate (Synapse, Google, RAS, ...) to the access token for later use.  Since access tokens are JSON Web Tokens (JWTs), which have data in a JSON payload, we add the IdP as a new field in the JSON.

There are a number of ways to generate an access token , e.g., logging in with a password, logging in using an OAuth provider (Google, RAS, ORCiD...), generating a personal access token (PAT).  This PR covers all paths, including some "edge cases" like admin services for generating an access token.

An OAuth client of Synapse can generate an access token, either through the "authorization code flow" or via a "refresh token" grant.  In both cases there is an *initial* request where we know what IdP was used and later a second request where we need to put the IdP into the access token. So we need "save" the IdP between the two requests.  To do this we associate the IdP with the one-time authorization code, and also with the refresh token.

In he case of the refresh token we store the IdP as a new field in the tokenmetadata.
In the case of the authorization code we store the IdO as a new field in the authorization request object.
2 factor auth is also a multi-step process.  we pass along the IdP in the twofactorauthtoken object.

Once the IdP is in the access token, we need to make it available for various services.  `AuthenticationFilter` gets the IdP from the access token and passes it along as an request header, along with other security related information.

### CODE REVIEW

File: services/repository/src/test/java/org/sagebionetworks/auth/filter/AuthenticationFilterTest.java [hunk 1]
```diff
@@ -217,6 +219,28 @@ public void testFilter_AccessTokenPassedAsSessionToken() throws Exception {
 		assertEquals(AuthenticationMethod.SESSIONTOKEN.name(), requestCaptor.getValue().getHeader(AuthorizationConstants.SYNAPSE_AUTHENTICATION_METHOD_HEADER_NAME));
 	}
 
+	@Test
```
Comment (john-hill) [inline services/repository/src/test/java/org/sagebionetworks/auth/filter/AuthenticationFilterTest.java:222]: This could be parameterized to test for each of the enums, advantage, always get to test new values.

File: lib/models/src/main/java/org/sagebionetworks/repo/model/auth/IdentityProviderUtils.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+public class IdentityProviderUtils {
+
+	/** The name of {@link SynapseIdentityProvider}, Synapse itself. */
+	public static final String SYNAPSE_IDENTITY_PROVIDER = "SYNAPSE";
+
+	/**
+	 * The name of the given identity provider, or null if there is none.
+	 */
+	public static String toName(IdentityProvider identityProvider) {
```
Comment (john-hill) [inline lib/models/src/main/java/org/sagebionetworks/repo/model/auth/IdentityProviderUtils.java:18]: If we have an enum that is the super-set of all of the types, this gets past at all levels.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/authentication/PersonalAccessTokenManagerImplUnitTest.java [hunk 3]
```diff
@@ -491,7 +491,7 @@ void testCreateToken_DuplicateKeyException() {
 		// Method under test
 		assertThrows(
 				IllegalArgumentException.class,
-				() -> personalAccessTokenManager.issueToken(userInfo, ACCESS_TOKEN, new AccessTokenGenerationRequest(), OAUTH_ENDPOINT),
+				() -> personalAccessTokenManager.issueToken(userInfo, ACCESS_TOKEN, new AccessTokenGenerationRequest(), OAUTH_ENDPOINT, null),
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/authentication/PersonalAccessTokenManagerImplUnitTest.java:494]: need a test for the non null case

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/service/AdministrationServiceImpl.java [hunk 4]
```diff
@@ -208,7 +209,8 @@ public LoginResponse createOrGetTestUser(Long userId, NewIntegrationTestUser use
 					.setState(VerificationStateEnum.APPROVED));
 		}
 		
-		return authManager.loginWithNoPasswordCheck(createdUser.getId(), null);
+		// The user is created with a password in Synapse, so Synapse is the provider
+		return authManager.loginWithNoPasswordCheck(createdUser.getId(), null, new SynapseIdentityProvider());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/service/AdministrationServiceImpl.java:213]: we can use the passed provider

File: lib/jdomodels/src/main/resources/schema/OAuthRefreshToken-ddl.sql [hunk 5]
```diff
@@ -6,6 +6,7 @@ CREATE TABLE IF NOT EXISTS `OAUTH_REFRESH_TOKEN` (
   `CLIENT_ID` BIGINT NOT NULL,
   `SCOPES_JSON` JSON NOT NULL,
   `CLAIMS_JSON` JSON NOT NULL,
+  `IDENTITY_PROVIDER` VARCHAR(64) DEFAULT NULL,
```
Comment (john-hill) [inline lib/jdomodels/src/main/resources/schema/OAuthRefreshToken-ddl.sql:9]: Enum

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/oauth/OAuthRefreshTokenInformation.json [hunk 6]
```diff
@@ -53,6 +53,16 @@
 			"description": "OIDC Claims that this refresh token grants access to"
 		},
 
+		"identityProvider": {
+			"type": "string",
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/oauth/OAuthRefreshTokenInformation.json:57]: can this be an enum too?

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/auth/TwoFactorAuthToken.json [hunk 7]
```diff
@@ -12,6 +12,10 @@
 		},
 		"context": {
 			"$ref": "org.sagebionetworks.repo.model.auth.TwoFactorAuthTokenContext"
+		},
+		"identityProvider": {
+			"$ref": "org.sagebionetworks.repo.model.auth.IdentityProvider",
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/auth/TwoFactorAuthToken.json:17]: enum instead of the interface

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/oauth/OIDCAuthorizationRequest.json [hunk 8]
```diff
@@ -37,6 +37,10 @@
 			"type": "string",
 			"format": "date-time",
 			"description": "Timestamp marking when authorization was approved (set by server)"
+		},
+		"identityProvider": {
+			"type": "string",
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/oauth/OIDCAuthorizationRequest.json:42]: be emumed!

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/authentication/AuthenticationManagerImpl.java [hunk 9]
```diff
@@ -191,7 +194,8 @@ void validateTwoFactorRequirementForPasswordChange(Long principalId) {
 		// See https://sagebionetworks.jira.com/browse/PLFM-8273, when the user updates the user password we need to make sure that
 		// the 2nd factor is used to perform the operations if enabled.
 		if (user.hasTwoFactorAuthEnabled()) {
-			throw new TwoFactorAuthRequiredException(user.getId(), twoFaManager.generate2FaToken(user, TwoFactorAuthTokenContext.PASSWORD_CHANGE));
+			throw new TwoFactorAuthRequiredException(user.getId(),
+					twoFaManager.generate2FaToken(user, TwoFactorAuthTokenContext.PASSWORD_CHANGE, null));
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/authentication/AuthenticationManagerImpl.java:198]: this should be the provider?

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/authentication/AuthenticationManagerImpl.java [hunk 10]
```diff
@@ -226,42 +230,50 @@ public LoginResponse login(LoginRequest request, String tokenIssuer){
...
 		
-		return getLoginResponseAfterSuccessfulAuthentication(user, issuer);
+		return getLoginResponseAfterSuccessfulAuthentication(user, issuer, identityProvider);
 	}
 	
 	@Override
 	public LoginResponse loginWithNoPasswordOrTwoFaCheck(UserInfo user, String issuer) {
-		return getLoginResponseAfterSuccessfulAuthentication(user, issuer);
+		// Only an administrator reaches this, and it is Synapse that authorized them to do so.
+		return getLoginResponseAfterSuccessfulAuthentication(user, issuer, new SynapseIdentityProvider());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/authentication/AuthenticationManagerImpl.java:258]: if we pass this in we can allow tests of different values.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/authentication/AuthenticationManagerImplUnitTest.java [hunk 11]
```diff
@@ -928,7 +934,65 @@ public void testLoginWith2Fa() {
...
+			.setOtpCode("123456")
+			.setTwoFaToken("2faToken");
+
+		// Call under test
+		authManagerSpy.loginWith2Fa(loginRequest, issuer);
+
+		verify(authManagerSpy).getLoginResponseAfterSuccessfulAuthentication(userInfo, issuer, orcid);
+	}
+
+	@Test
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/authentication/AuthenticationManagerImplUnitTest.java:967]: missing a test for the case where we were passing null.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/authentication/TwoFactorAuthManagerImplUnitTest.java [hunk 12]
```diff
@@ -610,7 +610,7 @@ public void testGenerate2FaToken(TwoFactorAuthTokenContext context) {
 			.setExpiresOn(new Date(12345 + TwoFactorAuthManagerImpl.TWO_FA_TOKEN_DURATION_MINS * 60 * 1000));
 				
 		// Call under test
-		String result = manager.generate2FaToken(user, context);
+		String result = manager.generate2FaToken(user, context, null);
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/authentication/TwoFactorAuthManagerImplUnitTest.java:613]: need test where we encode the provider.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/OpenIDConnectManagerImplUnitTest.java [hunk 13]
```diff
@@ -1568,7 +1585,7 @@ public void testValidateAccessTokenWithNoRefreshTokenId() {
 		claims.setSubject(ppid);
 
 		// method under test
-		assertEquals(USER_ID, openIDConnectManagerImpl.validateAccessToken(token));
+		assertEquals(new ValidatedAccessToken(USER_ID, null), openIDConnectManagerImpl.validateAccessToken(token));
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/OpenIDConnectManagerImplUnitTest.java:1588]: can we added tests for the 4+? cases where we pass in a value that is not null?

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/OAuthRefreshTokenManagerAutowiredTest.java [hunk 14]
```diff
@@ -324,7 +324,7 @@ public void testAuditAndRevokeTokensForUserClientPair() throws Exception {
 
 	@Test
 	public void testIsRefreshTokenActive() {
-		OAuthRefreshTokenAndMetadata token = refreshTokenManager.createRefreshToken(user1, client1.getClient_id(), scopes, claims);
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/oauth/OAuthRefreshTokenManagerAutowiredTest.java:327]: some non null cases.

**Review by john-hill — CHANGES_REQUESTED** [review]
